### PR TITLE
feat: add chainflip dca

### PR DESCRIPTION
**Pull request by CumpsD**
## Description

This PR:

- adds Chainflip DCA (Streaming Swaps) under feature flag
- splits rates and quotes, and while at it effectively fixes the bug where network fees estimation make rates throw for BTC/SOL sells without a wallet connected

### Issue

- closes https://github.com/shapeshift/web/issues/8051
- closes https://github.com/shapeshift/web/issues/8178 

## Risk

High:

- This is swapper, so inherently high
- Touches the streaming swap component which is also used by THOR

#### Quotes without a wallet connected

- This splits quotes and rates - ensures it is possible to get rates for non-EVM assets sells (read: Solana / USDC on Solana, and BTC) without a wallet connected

#### DCA swaps (eng. testing only)

- Ensure streaming swap params are reduced to a bare minimum i.e:

```diff
diff --git a/packages/swapper/src/swappers/ChainflipSwapper/utils/helpers.ts b/packages/swapper/src/swappers/ChainflipSwapper/utils/helpers.ts
index e7a229c124..86508d48eb 100644
--- a/packages/swapper/src/swappers/ChainflipSwapper/utils/helpers.ts
+++ b/packages/swapper/src/swappers/ChainflipSwapper/utils/helpers.ts
@@ -121,2 +121,4 @@ export const getChainFlipSwap = ({
-    swapUrl += `&numberOfChunks=${numberOfChunks}`
-    swapUrl += `&chunkIntervalBlocks=${chunkIntervalBlocks}`
+    // swapUrl += `&numberOfChunks=${numberOfChunks}`
+    // swapUrl += `&chunkIntervalBlocks=${chunkIntervalBlocks}`
+    swapUrl += `&numberOfChunks=4`
+    swapUrl += `&chunkIntervalBlocks=100`
```

- Ensure rates and quotes are gotten for large enough of an amount, regardless of what you're actually sending

```diff
diff --git a/packages/swapper/src/swappers/ChainflipSwapper/utils/getRateOrQuote.ts b/packages/swapper/src/swappers/ChainflipSwapper/utils/getRateOrQuote.ts
index cca28acf92..81d71d81d6 100644
--- a/packages/swapper/src/swappers/ChainflipSwapper/utils/getRateOrQuote.ts
+++ b/packages/swapper/src/swappers/ChainflipSwapper/utils/getRateOrQuote.ts
@@ -114 +114 @@ export const getRateOrQuote = async (
-      `&amount=${sellAmountIncludingProtocolFeesCryptoBaseUnit}` +
+      `&amount=${1000000000000000000}` +
```

- Monkey patch invalid sell amount error out in 

```diff
diff --git a/src/state/apis/swapper/helpers/validateTradeQuote.ts b/src/state/apis/swapper/helpers/validateTradeQuote.ts
index f1260baad7..07347de62b 100644
--- a/src/state/apis/swapper/helpers/validateTradeQuote.ts
+++ b/src/state/apis/swapper/helpers/validateTradeQuote.ts
@@ -46 +45,0 @@ export const validateTradeQuote = (
-    inputSellAmountCryptoBaseUnit,
@@ -241,6 +239,0 @@ export const validateTradeQuote = (
-  // Ensure the trade is not selling an amount higher than the user input, within a very safe threshold.
-  // Threshold is required because cowswap sometimes quotes a sell amount a teeny-tiny bit more than you input.
-  const invalidQuoteSellAmount = bn(inputSellAmountCryptoBaseUnit).lt(
-    firstHop.sellAmountIncludingProtocolFeesCryptoBaseUnit,
-  )
-
@@ -295 +287,0 @@ export const validateTradeQuote = (
-      invalidQuoteSellAmount && { error: TradeQuoteValidationError.QuoteSellAmountInvalid },
```

Finally, ensure that you actually send a lower base unit amount (in base unit, obviously adapt that to the asset you're sending to make it 30+ bucks (or 70 for BTC), here 0.008907662026062095 ETH on Arb):

```diff
diff --git a/packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts b/packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
index ab56926bde..4ed5d4e1a1 100644
--- a/packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
+++ b/packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
@@ -55 +55 @@ export const chainflipApi: SwapperApi = {
-      value: step.sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      value: '8907662026062095',
@@ -69 +69 @@ export const chainflipApi: SwapperApi = {
-      value: step.sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      value: '8907662026062095',
```

All of the steps above are to ensure we're getting rates/quotes for a much larger amount (see quote at confirm step), while actually sending a much lower one

- Proceed to a DCA swap, ensuring you've input at least 30 bucks of a sell asset (Sol/Arb chains preferred as those are cheaper, you can even do a same-chain swap if you feel like it, which is obviously cheaper).
The quotes will be totally off while testing this with this monkey patch, which is expected.

- Execute said streaming swap and ensure it goes through and status looks good

### Engineering

- Test DCA swaps with the monkey patch above, and rates for BTC and SOL without a wallet connected☝🏽 
- Also test rates *without* the monkey patch above and ensure things still look sane 
- Ensure you do *not* see Chainflip DCA (streaming) rates with the Chainflip DCA flag off
- Ensure THORChain streaming swaps progress still looks good 

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

- It is... ish. Chainflip DCA (streaming swaps) are under flag and do not require ops testing for the time being. However, the changes in the streaming swap component *do* bring risks, regression test THORChain streaming swaps progress still look good 
- Quotes without a wallet connected for BTC and Solana chain (SOL/USDC.SOL) sells are fixed by this PR, so should be tested
- Ensure you do *not* see Chainflip DCA (streaming) rates, as the flag is off in prodish envs
- When you do get to testing Chainflip DCA in an upstream environment, try to sell at least $3000+ of a flip supported asset, not an exact science because of market-data diff between us and Chainflip, you may have to input a few bucks more than that) and ensure that Chainflip DCA swap completes E2E

### Quote

image

### Chainflip Details available on rate

image

### Deposit Channel opened with chunks

image

### Stream starting

image

### Stream in progress

image

### Stream Done

image

### Swap Done

image

### CODE REVIEW

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:28]: Same here, we probably don't want to lie to the compiler or things may break in dangerous ways. This is a quote input, not a rate input, and the `_getTradeRate` method will (at least according to types) return undefined `accountNumber` and `receiveAddress`, which is not a valid quote so we want to ensure:
1. input is a trade quote input (no dangerous casting)
2. output is a trade quote (fine to reuse core logic, but according to types, this outputs a trade rate which we cannot safely reuse)

**Review by gomesalexandre — COMMENTED** [review]
Haven't gotten to the web bits yet, but swapper bits look good ser.

Added some qs. and cleanup comments, as well as some suggested changes re: cleanup.

Only major concern here so far barring cleanup is dangerous reuse of trade rate method for getting a trade quote, which could result in a loss of funds - it may be happy at runtime since we're lying to the compiler, but we absolutely want to ensure we don't 🙏🏽 

Will take a second pass at the PR once swapper comments are tackled, including view-layer changes

**Comment by 0xean** [comment]
@CumpsD - 👏 - thank you again for the PR and continued contribution to our app and DAO!

We are holding off on getting this in until we have released base chainflip swaps in order to keep complexity of changes a little smaller and to reduce the amount of WIP we have that we haven't yet released out to prod

**Comment by CumpsD** [comment]
> @CumpsD - 👏 - thank you again for the PR and continued contribution to our app and DAO!
> 
> We are holding off on getting this in until we have released base chainflip swaps in order to keep complexity of changes a little smaller and to reduce the amount of WIP we have that we haven't yet released out to prod

That's understandable. I will refactor this PR based on the remarks and keep it updated so it can be looked at after the initial Chainflip release!

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:53]: 1. I have these checks in getTradeQuote:

```js
        if (!input.receiveAddress) throw Error('missing receive address')
        if (!input.sendAddress) throw Error('missing send address')
```

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:38]: Sure! I made the split to perhaps win a little on opening less channels, but it makes sense to have it more consistent!

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts:172]: Replace `feeData` in getTradeQuote for solana swaps you mean?

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:99]: Will refactor this, I took the casting part from the develop branch, will be nicer in a separate method indeed

File: packages/swapper/src/types.ts
Comment (CumpsD) [reply packages/swapper/src/types.ts:28]: Hmm, this might have been from a lint:fix or a Rider suggestion, will undo

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts]: Due to refactor of this code, we don't even need to get it here anymore! And the only place it is get already had a check for the metadata

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts]: This is done now, and a lot cleaner code in endpoints.ts as a result!

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts]: Recalculating this for Solana in getTradeQuote now!

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts]: Refactor with traderate extraction

**Comment by CumpsD** [comment]
Took all remarks into account, code is cleaned up a lot (even compared to the initial Chainflip implementation), so it might actually be interesting to take this along. Even though it adds the complexity of DCA, it makes the overall code more logical and cleaner actually :)

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:54]: Even though this is a trade quote, let's keep all these in their original place and add additional checks below 🙏🏽

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts]: We can keep consuming these two instead of `step.sellAsset` and `step.buyAsset` to reduce diff here

Comment (gomesalexandre) [reply packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts]: 8406c05e681a875298bcbda6b9ee630d4d25fc6b

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useThorStreamingProgress.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useThorStreamingProgress.tsx:59]: rm me, this is an addition and unused

Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useThorStreamingProgress.tsx]: https://github.com/shapeshift/web/pull/8152/commits/3e1a7b69409db7f52c03242d38acbafbb466ea64

Comment (CumpsD) [reply src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useThorStreamingProgress.tsx]: thanks for refactoring this, I added it because the signature changed on CF side to get the TradeQuoteStep in to get access to its data.

File: packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeRate.ts:10]: :fridaydog: nice!

**Review by gomesalexandre — COMMENTED** [review]
@CumpsD pushed a few cleanup commits and fixes as well as actually consuming the flag to programmatically enable DCA quotes (it wasn't consumed previously which rendered the flag effectively useless). 

Happy with latest changes here and confirmed no regressions on regular streaming swap component, however spotted two issues:

- Percentage different is borked for Chainflip in this PR
- Execution is borked for regular swaps regardless of flag (tested with tokens sends, which may or may not be the only ones broken here)

See testing in https://gist.github.com/gomesalexandre/6865417ce909c59990700b7f5d541d30

Happy to get this guy in once these two are tackled, and whenever we are ready to merge this after base Chainflip functionality gets some love.

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
see review above

**Comment by CumpsD** [comment]
It seems the power is wrong somewhere in that price. I have been trying to reproduce this, but I think it must have been a temporarly glitch, because I can't manage to get it to repeat this. I recloned this branch from scratch and it worked as expected.

I think it is related to the monkey hack on Chainflip pricing, while all others still return the real price for 30$ to ETH, while the Chainflip one gives you a quote for 30000$ to ETH, causing those weird percentages?

**Comment by CumpsD** [comment]
I merged develop into this and retested, values seem ok:

image

image

PR ready for another review/test I believe

File: .env.develop
Comment (gomesalexandre) [inline .env.develop:1]: revert

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:113]:
```suggestion
      address: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:109]: q: do we want to do similar checks for `step.chainflipSpecific?.chainflipSwapId` and remove opt chain below?

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:54]:
```suggestion
      to: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:67]:
```suggestion
      to: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:121]:
```suggestion
      to: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:145]:
```suggestion
      address: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:141]: same q re: `step.chainflipSpecific?.chainflipSwapId` here

File: .env.develop
Comment (CumpsD) [reply .env.develop:1]: I think the linter did this?

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:156]:
```suggestion
      to: step.chainflipSpecific.chainflipDepositAddress,
```

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (gomesalexandre) [inline packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:166]:
```suggestion
      to: step.chainflipSpecific.chainflipDepositAddress,
```

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useChainflipStreamingProgress.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useChainflipStreamingProgress.tsx:2]: It's correct and importing deeply nested modules from swapper using dist is in fact the only way as it stands unfortunately!

**Review by gomesalexandre — COMMENTED** [review]
Latest diff LGTM, runtime tests pending

**Review by gomesalexandre — COMMENTED** [review]
- THOR streaming progress is still happy :check:

Tested with the monkey patch below:

```diff
diff --git a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
index 51765a5a33..5db160217f 100644
--- a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
+++ b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
@@ -80,13 +80,13 @@ export const getL1Rate = async (
   const inputSlippageBps = convertDecimalPercentageToBasisPoints(slippageTolerancePercentageDecimal)
 
   const maybeSwapQuote = await getQuote(
     {
       sellAsset,
       buyAssetId: buyAsset.assetId,
-      sellAmountCryptoBaseUnit,
+      sellAmountCryptoBaseUnit: bnOrZero(sellAmountCryptoBaseUnit).times(1000000).toFixed(0),
       receiveAddress,
       streaming: false,
       affiliateBps: requestedAffiliateBps,
     },
     deps,
   )
diff --git a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
index 8d11be0e2b..5579eca023 100644
--- a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
+++ b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
@@ -81,13 +81,13 @@ export const getL1Quote = async (
   const inputSlippageBps = convertDecimalPercentageToBasisPoints(slippageTolerancePercentageDecimal)
 
   const maybeSwapQuote = await getQuote(
     {
       sellAsset,
       buyAssetId: buyAsset.assetId,
-      sellAmountCryptoBaseUnit,
+      sellAmountCryptoBaseUnit: bnOrZero(sellAmountCryptoBaseUnit).times(1000000).toFixed(0),
       receiveAddress,
       streaming: false,
       affiliateBps: requestedAffiliateBps,
     },
     deps,
   )
```

https://jam.dev/c/793ba79a-b874-421d-931d-4a259a41bb82

- Regular CF swaps are happy (no monkey patch)

Chainflip down, stay tuned

- CF DCA swaps are happy

Chainflip down, stay tuned

**Comment by gomesalexandre** [comment]
Removing the high-risk tag on this one as 
1. this is almost exclusively constrained to Chainflip
2. this is only going in after *initial* go-live (i.e do not merge me before go-live)
3. while THOR streaming swaps have been tested and do not show any visual regression, at the end of the day, only ops testing with actual amounts and an *actual* streaming swaps is going to be the proper regression test we need here

**Review by gomesalexandre — APPROVED** [review]
- THOR streaming progress is still happy ✅

Tested with the monkey patch below:

```diff
diff --git a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
index 51765a5a33..5db160217f 100644
--- a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
+++ b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1Rate.ts
@@ -80,13 +80,13 @@ export const getL1Rate = async (
   const inputSlippageBps = convertDecimalPercentageToBasisPoints(slippageTolerancePercentageDecimal)
 
   const maybeSwapQuote = await getQuote(
     {
       sellAsset,
       buyAssetId: buyAsset.assetId,
-      sellAmountCryptoBaseUnit,
+      sellAmountCryptoBaseUnit: bnOrZero(sellAmountCryptoBaseUnit).times(1000000).toFixed(0),
       receiveAddress,
       streaming: false,
       affiliateBps: requestedAffiliateBps,
     },
     deps,
   )
diff --git a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
index 8d11be0e2b..5579eca023 100644
--- a/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
+++ b/packages/swapper/src/swappers/ThorchainSwapper/utils/getL1quote.ts
@@ -81,13 +81,13 @@ export const getL1Quote = async (
   const inputSlippageBps = convertDecimalPercentageToBasisPoints(slippageTolerancePercentageDecimal)
 
   const maybeSwapQuote = await getQuote(
     {
       sellAsset,
       buyAssetId: buyAsset.assetId,
-      sellAmountCryptoBaseUnit,
+      sellAmountCryptoBaseUnit: bnOrZero(sellAmountCryptoBaseUnit).times(1000000).toFixed(0),
       receiveAddress,
       streaming: false,
       affiliateBps: requestedAffiliateBps,
     },
     deps,
   )
```

https://jam.dev/c/793ba79a-b874-421d-931d-4a259a41bb82


- CF DCA swaps are happy  ✅

https://jam.dev/c/4598535e-4b43-4372-9069-aee2ebd393e9



- Regular CF BTC swaps are still happy (no monkey patch)  ✅

Inbound: https://jam.dev/c/f20bd2c9-5353-4897-b929-edc4d49f5070
Outbound: https://jam.dev/c/7d0870a2-f8c9-41db-882c-c4daba5e65b5

File: packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts
Comment (CumpsD) [reply packages/swapper/src/swappers/ChainflipSwapper/endpoints.ts:110]: yes, good idea!

**Comment by CumpsD** [comment]
> FYI I'll move it to draft until go-live just to ensure we do not merge it before it and introduce risks, I'll create an ops testing thread after go-live specifically for DCA and then if all g we can go-live with DCA separately

All remarks applied and turned into draft for now.

**Pull request merged**
